StudyNote API 리팩터링 및 test코드 추가 등

monda95 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #123
- 작업 요약: StudyNote 리팩터링

## 📄 상세 내용
- NoteAPI : premission =Allowany (테스트 간단화), 
- permissions : 멤버확인으로 생성하며 작성자 권한으로 수정 및 제거  
- testcode 작성
- url mapping : GET 은 그룹uuid 종속된 엔드포인트 POST와 detail은 note, note/int:id
- AI요약을 DetailView에 통합시켜서 service/notes에서 호출 (현재는 mock상태)

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Meoyoug commented on .vscode/launch.json:
컥..,, 다몬님 vscode 쓰시나보네요 

.vscode 디렉터리 .gitignore에 추가하시고 관련 파일들 커밋에서 제거하고 다시 푸시 부탁드려용

Meoyoug commented on .vscode/settings.json:
이것도!

2joonkim commented on .vscode/launch.json:
```diff
@@ -0,0 +1,15 @@
+{
```
다몬님 현재 vscode로 작업을 하신걸로 보입니다! 
.vscode같은 ide 캐시파일은 gitignore에 넣어주셔야 합니다!
(그래도 된다면 꼭 pycharm으로 작업해주세요!!)

2joonkim commented:
현재 코드 보기에 앞서 커밋 상황들이 꼬인것으로 보입니다!
나중에 같이 보시면서 해결 진행하고 pr 올리시는걸로 진행하시죠

monda95 committed:
✨ note모델 마이그레이션 및 그룹 모델 수정

- 그룹 UUIDbaseModel 상속하게 수정
- note model 마이그레이션 마이그레이트

# Conflicts:
#	apps/studies/models/groups.py

# Conflicts:
#	apps/studies/models/__init__.py

monda95 committed:
✨ Note API 리팩터링 및 URL 매핑 수정

- Note API Allowany 및 간략화
- AI 요약 로직 serviece.notes에 이동해서 불러오기
- url mapping group에 종속되게끔 변경(요구사항도 대로 구현)

# Conflicts:
#	apps/studies/urls.py

# Conflicts:
#	apps/studies/urls.py

monda95 committed:
fix : 머지 이후에 url값 상이해져 다시 testcode수정

2joonkim commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,186 @@
+    @extend_schema(summary="스터디 노트 생성 API")
+    def post(self, request: Request, *args: Any, **kwargs: Any) -> Response:
+        group_id = kwargs.get("group_id")
+        group = get_object_or_404(StudyGroup, uuid=self.kwargs["group_id"])
```
이 부분 기능적으로는 문제 없지만 코드에서 같은 값을 두번 가져오고 있습니다
그래서 통일해주는게 좋아보입니다!

```
group_uuid = kwargs["group_id"]
group = get_object_or_404(StudyGroup, uuid=group_uuid)
```

2joonkim commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,186 @@
+        group_id = kwargs.get("group_id")
+        note_id = kwargs.get("note_id")
+
+        note = get_object_or_404(StudyNote, id=note_id, study_group_id=group_id)
```
요부분 다른 api들과는 다른데 혹시 이유가 있을까요?
타 api와들과는 다르게 설정해두신 거 같은데 이렇게 설정해두셨을때
group_id 타입이 안 맞을 수 도 있을 거 같습니다

Meoyoug commented on apps/studies/models/groups.py:
```diff
@@ -13,7 +13,9 @@ class StudyGroupStatus(models.TextChoices):
 
 
-class StudyGroup(BaseModel):
+class StudyGroup(UUIDBaseModel):
```
누락된 필드 추가 감사합니당

Meoyoug commented on apps/studies/serializers/notes.py:
```diff
@@ -0,0 +1,93 @@
+User = get_user_model()
+
+
+class AuthorSlimSerializer(serializers.ModelSerializer[Any]):
```
StudyNoteAuthorSerializer 라는 명확한 명칭어때요?!

Meoyoug commented on apps/studies/serializers/notes.py:
```diff
@@ -0,0 +1,93 @@
+            raise serializers.ValidationError({"title": "필수 항목입니다."})
+        if not attrs.get("content"):
+            raise serializers.ValidationError({"content": "필수 항목입니다."})
+        return attrs
```
이거 title, content에 대해서 required=False 안해주셨으면 기본값이라 해당 필드에 대한 필수 여부 검증이 모델 시리얼라이저에 의해서 자동으로 이루어질거에요!
해당 함수는 삭제해도 무방합니당

Meoyoug commented on apps/studies/serializers/notes.py:
```diff
@@ -0,0 +1,93 @@
+    def validate(self, attrs: Dict[str, Any]) -> Dict[str, Any]:
+        if not attrs:
+            raise serializers.ValidationError({"error": "title 또는 content 중 하나는 필요합니다."})
+        return attrs
```
만약 이 시리얼라이저가 부분 수정인 PATCH HTTP 메소드 요청에 사용할 거라면 extra_kwargs 에서 required 속성을 추가하지 않아도되고 serializer = StudyNoteUpdateSerializer(request.data, patial=True)로 하시면됩니다.

그럼 Validate 함수도 없어도 될거에요

Meoyoug commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,186 @@
+        )
+
+
+class StudyNoteSummaryAPIView(APIView):
```
요약만 따로 조회해야하는 이유가 있을까요?
StudyNote 상세 조회에서 같이 줘도될거같은데

Meoyoug commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,186 @@
+        note = get_object_or_404(StudyNote, id=note_id, study_group=group)
+
+        if note.author != request.user:
+            raise PermissionDenied("해당 노트를 삭제할 권한이 없습니다.")
```
이부분은 CustomPermission으로 permissions.py에 별도로 정의부탁드려요~

Meoyoug commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,186 @@
+    @extend_schema(summary="스터디 노트 삭제 API")
+    def delete(self, request: Request, *args: Any, **kwargs: Any) -> Response:
+        group_uuid = kwargs.get("group_id")
+        note_id = kwargs.get("note_id")
```
필요한 인자가 정해져있다면 *args, **kwargs 가 아닌 group_uuid, note_id를 파라미터로 path param의 순서에 맞게 추가해주시는게 좋아요

Meoyoug commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,186 @@
+        note_id = kwargs.get("note_id")
+
+        group = get_object_or_404(StudyGroup, uuid=group_uuid)
+        note = get_object_or_404(StudyNote, id=note_id, study_group=group)
```
노트만 조회하고 노트와 FK로 엮인 Group을 Join으로 가져와서 Permission을 검증하는 방법은 어떨까요?
실제로 요청 처리에 필요한 객체로 Note 객체 하나인거같은데요?

Meoyoug commented on apps/studies/urls.py:
```diff
@@ -15,4 +20,13 @@
+    path("groups/<uuid:group_id>/notes/", StudyNoteListCreateAPIView.as_view(), name="study-note-list-create"),
+    path("groups/<uuid:group_id>/notes/<int:note_id>/", StudyNoteDetailAPIView.as_view(), name="study-note-detail"),
+    path(
+        "groups/<uuid:group_id>/notes/<int:note_id>/summary/",
```
굳이 Resource의 Depth를 나타내는 형태로 url을 구성하지 않으셔도 됩니다!

현재 구현하신 파트가 StudyNote 에 대한 엔드포인트밖에없는 것 같아요

study-notes/
study-notes/<int:note_id>

이두개면 충분해보여요

monda95 replied:
피드백 너무 감사합니다 일단 최대한 다 수렴해서 수정해봤는데, 확인부탁드림다

monda95 committed:
fix : studies.models.__init__  isort

-from .groups import StudyGroup
from .notes import StudyNote
from .reviews import Review 로 순서변경

monda95 committed:
add : .gitignore

팀원 브랜치 Pull하고 rebase하는 과정에서 vscode 잔재가 딸려옴. .vscode 추가

monda95 committed:
fix : reviewer feedback 수렴

- AI 요약 로직 통합 (StudyNoteService.summarize() 추가, AIService 제거)
- UUID 기반 구조 → FK 기반 단순화 및 URL Depth 축소 (/study-notes/)
- Summary API 제거 및 Detail 응답에 요약 통합
- 퍼미션 구조 정비 (IsGroupMember 신설, views는 self.check_object_permissions() 적용하되 AllowAny로 임시 허용)
- Serializer/Service/View 구조 통합 및 테스트 코드 FK 기반으로 수정
- models/__init__.py에 __all__ 추가

Meoyoug commented on apps/studies/permissions.py:
```diff
@@ -10,3 +14,53 @@ class IsGroupLeader(BasePermission):
+
+        # 인증되지 않은 경우 읽기(GET)만 허용
+        if not user.is_authenticated:
+            return request.method == "GET"
```
스터디 노트는 해당 그룹에 속한 멤버끼리 공유가 가능한 기능으로 해당 절은 False로 리턴해주셔야 합니다

Meoyoug commented on apps/studies/permissions.py:
```diff
@@ -10,3 +14,53 @@ class IsGroupLeader(BasePermission):
+            return is_member
+
+        elif request.method in ("PUT", "PATCH", "DELETE"):
+            return is_member and is_author
```
수정과 삭제는 작성자만 되게 하려면 IsStudyNoteAuthor  라는 역할로 분리하는게 더 적절할 것 같습니다.

하나의 Permission클래스는 하나의 권한에 대해서만 동작하게 설계하시도록 하는 것이 좋습니다.

Meoyoug commented on apps/studies/permissions.py:
```diff
@@ -10,3 +14,53 @@ class IsGroupLeader(BasePermission):
+            return is_member or user.is_authenticated
+
+        elif request.method == "POST":
+            return is_member
```
위에서 if not user.is_authenticated 일때 return False를 해주니까 method 구분절은 다 지우고 마지막에 return is_member 만 해주시면 될 것 같아요

Meoyoug commented on apps/studies/permissions.py:
IsGroupMember, IsStudyNoteAuthor 이렇게 두개의 Permission으로 분리부탁드립니다!

Meoyoug commented on apps/studies/services/notes.py:
```diff
@@ -0,0 +1,82 @@
+
+    @staticmethod
+    def delete_mock(note_id: int) -> None:
+        return None
```
현재 View 에서나 Test에서나 해당 메서드들을 사용하는 부분이 없어보입니다!
이미 View 작성도 Spec 이나 Mock이 아닌 기능이 동작하도록 작성하셨기 때문에 해당 메서드들은 지워도 될 것 같아요

monda95 committed:
fix : reviewer feedback 수렴 (2)

- permissions 멤버 확인과, 수정 삭제 권한 역할별로 분리 (IsGroupMember / IsStudyNoteAuthor)
- services/notes.py 내 mock 메서드 제거 (AI 요약 로직만 유지)

2joonkim commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,146 @@
+    @extend_schema(summary="스터디 노트 생성 API")
+    def post(self, request: Request) -> Response:
+        serializer = StudyNoteCreateSerializer(data=request.data)
+        if not serializer.is_valid():
```
요 부분 drf의 벨리데이션 에러를 활용하신다면 그냥 에러메세지만 짧게 커스텀이 가능할거 같아요!
```
from rest_framework.exceptions import ValidationError

if not serializer.is_valid():
      rasie ValidationError("유효성 검사에 실패했습니다.")
```

2joonkim replied:
아래의 예외발생 케이스들도 동일합니다!

2joonkim replied:
이유 확인했습니다!!

2joonkim reviewed (changes requested):
코치님 피드백 받고 머지 진행하시지요!

2joonkim commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,146 @@
+from apps.studies.services.notes import StudyNoteService
+
+
+class StudyNoteListCreateAPIView(APIView):
```
아래의 url 부분의 리뷰를 확인해주시고 이 부분을 srp 따라서 변경해주시면 좋을 거 같습니다!
이 부분을 생성 전용으로 축소시키고 아래에 적용시켜주시면 좋을 거 같습니다!

2joonkim commented on apps/studies/urls.py:
```diff
@@ -37,4 +41,7 @@
     path("groups/", StudyGroupListCreateView.as_view(), name="study-group-list-create"),
     path("groups/<uuid:group_id>/", StudyGroupDetailUpdateView.as_view(), name="study-group-detail-update"),
+    # StudyNote APIs
+    path("notes/", StudyNoteListCreateAPIView.as_view(), name="study-note-list-create"),
```
현재 urls.py에서 StudyNoteListCreateAPIView를 notes/ 경로로 연결하고 있습니다!
그런데 StudyNoteListCreateAPIView 내부에는 유저 식별자(user_id)나 그룹 식별자(group_id)를 조회할 수 있는 로직이 존재하지 않습니다.

그래서 전달 구조를 확인하시어 한가지 방법으로 변경하시는걸 권장드립니다!

Meoyoug replied:
노트 목록 조회할 때는
api/v1/groups/<uuid:group_id>/notes

노트 생성할때는 
api/v1/notes

엔드포인트가 이렇게 정리되면 좋을 것 같아요

2joonkim commented on apps/studies/urls.py:
```diff
@@ -37,4 +41,7 @@
     ),
```
예시입니다 참고만 해주세요
```
urlpatterns = [
    path(
        "groups/<uuid:group_id>/",
        StudyGroupDetailUpdateView.as_view(),
        name="study-group-detail-update",
    ),
    path(
        "notes/<uuid:user_id>/",
        StudyNoteCreateAPIView.as_view(),
        name="study-note-create",
    ),
    path(
        "notes/<int:note_id>/",
        StudyNoteDetailAPIView.as_view(),
        name="study-note-detail",
    ),
]
```

monda95 committed:
fix : group_uuid로직 추가, 퍼미션 강화, 리스트 files_count 추가

- POST /notes/ 바디에 group_uuid 추가(writeOnly) → 퍼미션(IsGroupMember.has_permission)에서 형식/존재/멤버십 검증 후 view._group 주입 → 시리얼라이저(StudyNoteCreateSerializer.validate)가 이를 study_group FK로 매핑, author=CurrentUserDefault() 자동 주입 → 뷰는 is_valid(); save()만 수행.

- list GET은 `annotate(Count("attachments"), Count("images"))`로 갯수 카운트 한 후 `ExpressionWrapper(F(...) + F(...))`로 **files_count** 로 합산, 리스트 시리얼라이저에 `read_only`로 노출.

-그룹 선택·작성자 주입을 뷰/모델이 아닌 시리얼라이저 validate/create로 이동(뷰는 얇게 유지). 조회 시 상태 변경(요약 호출) 제거.

- url 유지하며 생성은 body의 group_uuid로 그룹식별

monda95 committed:
fix : isorted

- poetry install 후 isort

monda95 committed:
fix : 마이그레이션 fake로 맞춤 + test 코드 수정

Meoyoug commented on apps/studies/views/notes.py:
```diff
@@ -0,0 +1,144 @@
+                    F("attachment_count") + F("image_count"),
+                    output_field=IntegerField(),
+                )
+            )
```
<img width="381" height="228" alt="Image" src="https://github.com/user-attachments/assets/cd98542d-35dc-4bda-a1ad-d130ea07aa58" />

이부분 때문에 아마 파일카운트를 추가하신것같은데 첨부 파일에 대한 카운트만 반환하도록해주세요
image는 마크다운 에디터에 삽입된 이미지로 첨부파일과 별개입니다!

monda95 committed:
fix : GET, POST 엔드포인트 분리 및 파일만 카운트

- view 분리 + url 엔드포인트 분리 (그룹 uuid에 종속되는 노트 GET)

- 이미지파일 + 첨부파일로 갯수를 산출하던 로직을 annotate로 파일 갯수만 GET하도록 수정

monda95 committed:
fix : test 코드 수정

- test 코드도 알맞게 분기하여 수정

- CreatAPIView에 퍼미션, 파서 빠진 부분 추가

monda95 committed:
fix : blacked

monda95 committed:
✨feat : StudyNote, StudyNoteImage, StudyNoteAttachment 모델 마이그레이션 수행

Meoyoug reviewed (dismissed):
다몬님 고생많으셨습니다!

마이그레이션 파일 부분은 현재 개발 서버랑 충돌이 있을것 같아서 제가 커밋수정한뒤에 강제푸시해서 해결해뒀습니다!
